rotkehlchen/chain/ethereum/uniswap/uniswap.py

```diff
@@ -311,6 +408,198 @@ def _tx_swaps_to_trades(swaps: List[AMMSwap]) -> List[AMMTrade]:
+        min_end_ts: Timestamp = to_timestamp
+
+        # Get addresses' last used query range for Uniswap events
+        for address in addresses:
+            entry_name = f'{UNISWAP_EVENTS_PREFIX}_{address}'
+            events_range = self.database.get_used_query_range(name=entry_name)
+
+            if not events_range:
+                new_addresses.append(address)
+            else:
+                existing_addresses.append(address)
+                min_end_ts = min(min_end_ts, events_range[1])
+
+        # Request new addresses' events
+        if new_addresses:
+            start_ts = Timestamp(0)
+            for address in new_addresses:
+                for event_type in EventType:
+                    new_address_events = self._get_events_graph(
+                        address=address,
+                        start_ts=start_ts,
+                        end_ts=to_timestamp,
+                        event_type=event_type,
+                    )
+                    if new_address_events:
+                        address_events[address].extend(new_address_events)
+
+                # Insert new address' last used query range
+                self.database.update_used_query_range(
+                    name=f'{UNISWAP_EVENTS_PREFIX}_{address}',
+                    start_ts=start_ts,
+                    end_ts=to_timestamp,
+                )
+
+        # Request existing DB addresses' events
+        if existing_addresses and min_end_ts <= to_timestamp:
+            for address in existing_addresses:
+                for event_type in EventType:
+                    address_new_events = self._get_events_graph(
+                        address=address,
+                        start_ts=min_end_ts,
+                        end_ts=to_timestamp,
+                        event_type=event_type,
+                    )
+                    if address_new_events:
+                        address_events[address].extend(address_new_events)
+
+                # Update existing address' last used query range
+                self.database.update_used_query_range(
+                    name=f'{UNISWAP_EVENTS_PREFIX}_{address}',
+                    start_ts=min_end_ts,
+                    end_ts=to_timestamp,
+                )
+
+        # Insert requested events in DB
+        all_events = []
+        for address in filter(lambda address: address in address_events, addresses):
+            all_events.extend(address_events[address])
+
+        self.database.add_uniswap_events(all_events)
+
+        # Fetch all DB events within the time range
+        for address in addresses:
+            db_events = self.database.get_uniswap_events(
+                from_ts=from_timestamp,
+                to_ts=to_timestamp,
+                address=address,
+            )
+            if db_events:
+                # return events with the oldest first
+                db_events.sort(key=lambda event: (event.timestamp, event.log_index))
```

Comment:
shouldn't you put a `-` in front of `event.log_index` if you want to sort first by timestamp and then by log index?

Reply:
I've seen you have placed the `-` for the trades, but here my intention is to sort them in ascending mode (most recent last). Like this:
```python
a1 = {'a': 12345, 'b': 1}
a2 = {'a': 12346, 'b': 2}
a3 = {'a': 12346, 'b': 1}
a4 = {'a': 12340, 'b': 10}
a5 = {'a': 12340, 'b': 9}
a6 = {'a': 12340, 'b': 11}

l1 = [a6, a4, a5, a3, a1, a2]

 l1.sort(key=lambda x: (x['a'], x['b']))

# Expected l1 output
[{'a': 12340, 'b': 9},
 {'a': 12340, 'b': 10},
 {'a': 12340, 'b': 11},
 {'a': 12345, 'b': 1},
 {'a': 12346, 'b': 1},
 {'a': 12346, 'b': 2}]
```
In fact I haven't seen multiple events in the same tx, could it happen with a contract account tx?

Reply:
>  but here my intention is to sort them in ascending mode (most recent last)

Aha I see. Okay then. Thank you for the example.

> In fact I haven't seen multiple events in the same tx, could it happen with a contract account tx?

Yes definitely. That's why we have the log index. An ethereum transaction does something and has a transaction receipt, which may contain many many log entries. And each one of them has an index.

Example transaction and logs and their indices screenshot from etherscan:
https://etherscan.io/tx/0x95c31c24811aa89890f586455230a21b5e6805571291c41f2429c0b27ffa6494#eventlog

![2020-11-19_01-26](https://user-images.githubusercontent.com/1658405/99604657-4cb47f80-2a06-11eb-8231-6a7bbdbee8a0.png)